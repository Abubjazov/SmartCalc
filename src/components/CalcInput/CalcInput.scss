@import '../../variables';

.calc-input {
	.calc-input-warning {
		position: static;
		color: $red;
		font-size: 13px;
		line-height: 17px;
	}

	input {
		height: 25px;
		width: 260px;
		border: none;
		border-bottom: 1px solid $deep-gray;
		outline: none;
		background: none;
		font-size: 20px;
		font-weight: 300;

		&:focus {
			border-bottom: 2px solid $main-color;
		}
	}

	.warning {
		border-bottom: 1px solid $deep-red;
	}
}
