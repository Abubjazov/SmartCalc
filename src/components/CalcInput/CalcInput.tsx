import { useEffect, useState } from 'react'
import { useActions } from '../../hooks/useActions'

import './CalcInput.scss'

export const CalcInput = ({
	itemKey,
	itemValue,
	submitStatus,
}: {
	itemKey: string
	itemValue: string
	submitStatus: boolean | null
}): JSX.Element => {
	const { changeInputItem } = useActions()

	const [isValid, setIsValid] = useState<boolean | null>(null)

	const changeInput = (value: string) => {
		changeInputItem(itemKey, value.replace(/\D/, ''))
		setIsValid(value.length > 0)
	}

	useEffect(() => {
		submitStatus !== null && setIsValid(itemValue.length > 0)

		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [submitStatus])

	return (
		<div className='calc-input'>
			<input
				className={isValid !== null ? (isValid ? 'null' : 'warning') : 'null'}
				key={itemKey}
				type='text'
				placeholder='Введите число'
				onChange={e => changeInput(e.target.value)}
				onBlur={e => changeInput(e.target.value)}
				value={itemValue}
			/>

			{isValid !== null ? (
				isValid ? null : (
					<p className='calc-input-warning'>Введите число</p>
				)
			) : null}
		</div>
	)
}
