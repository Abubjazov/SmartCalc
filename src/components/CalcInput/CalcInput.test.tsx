import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import { CalcInput } from './CalcInput'

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

describe('Component: CalcInput', () => {
	test('should render CalcInput default', () => {
		const store = mockStore({})

		const { asFragment } = render(
			<Provider store={store}>
				<CalcInput itemKey='a13key1' itemValue='' submitStatus={null} />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()

		userEvent.tab()

		expect(asFragment()).toMatchSnapshot()

		userEvent.tab()

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render CalcInput input value, submitStatus is null', () => {
		const store = mockStore({
			calc: { inputItems: [{ key: 'a13key1', value: '' }] },
		})

		const { asFragment } = render(
			<Provider store={store}>
				<CalcInput itemKey='a13key1' itemValue='' submitStatus={null} />
			</Provider>
		)
		const inputItem = screen.getByPlaceholderText('Введите число')

		userEvent.type(inputItem, '1')

		expect(asFragment()).toMatchSnapshot()

		userEvent.type(inputItem, '13')

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render CalcInput input value, submitStatus is true', () => {
		const store = mockStore({
			calc: { inputItems: [{ key: 'a13key1', value: '' }] },
		})

		const { asFragment } = render(
			<Provider store={store}>
				<CalcInput itemKey='a13key1' itemValue='' submitStatus={true} />
			</Provider>
		)
		const inputItem = screen.getByPlaceholderText('Введите число')

		userEvent.type(inputItem, '1')

		expect(asFragment()).toMatchSnapshot()
	})
})
