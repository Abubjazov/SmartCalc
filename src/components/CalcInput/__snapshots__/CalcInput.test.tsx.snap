// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CalcInput should render CalcInput default 1`] = `
<DocumentFragment>
  <div
    class="calc-input"
  >
    <input
      class="null"
      placeholder="Введите число"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`Component: CalcInput should render CalcInput default 2`] = `
<DocumentFragment>
  <div
    class="calc-input"
  >
    <input
      class="null"
      placeholder="Введите число"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`Component: CalcInput should render CalcInput default 3`] = `
<DocumentFragment>
  <div
    class="calc-input"
  >
    <input
      class="warning"
      placeholder="Введите число"
      type="text"
      value=""
    />
    <p
      class="calc-input-warning"
    >
      Введите число
    </p>
  </div>
</DocumentFragment>
`;

exports[`Component: CalcInput should render CalcInput input value, submitStatus is null 1`] = `
<DocumentFragment>
  <div
    class="calc-input"
  >
    <input
      class="null"
      placeholder="Введите число"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`Component: CalcInput should render CalcInput input value, submitStatus is null 2`] = `
<DocumentFragment>
  <div
    class="calc-input"
  >
    <input
      class="null"
      placeholder="Введите число"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`Component: CalcInput should render CalcInput input value, submitStatus is true 1`] = `
<DocumentFragment>
  <div
    class="calc-input"
  >
    <input
      class="null"
      placeholder="Введите число"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
